feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go

```diff
@@ -0,0 +1,625 @@
+			wantWrite:  true,
+			wantRead:   true,
+			wantStatus: 0,
+		}, {
+			desc:       sDesc,
+			lowID:      inId90,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   false,
+			wantWrite:  false,
+			wantRead:   true,
+			wantStatus: 5,
+		}, {
+			desc:       pDesc,
+			lowID:      inId100,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   false,
+			wantWrite:  true,
+			wantRead:   true,
+			wantStatus: 0,
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.desc, func(t *testing.T) {
+			resp, err := streamP4RTArb(&test)
+			if err != nil && test.wantFail {
+				t.Errorf("Failed to setup P4RT Client: %v", err)
+			}
+			// Validate status code
+			if resp != test.wantStatus {
+				t.Errorf("Incorrect status code received: want %d, got %d", test.wantStatus, resp)
+			}
+			// Validate the response for Read/Write for each client
+			validateRWResp(t, &test)
+			// Disconnect Primary
+			removeClient(test.handle)
+		})
+	}
+}
+
+// Test Primary/Secondary clients
+func TestPrimarySecondary(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	configureDeviceId(t, dut)
+	configurePortId(t, dut)
+	testCases := []testArgs{
+		{
+			desc:       pDesc,
+			lowID:      inId100,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   false,
+			wantWrite:  true,
+			wantRead:   true,
+			wantStatus: 0,
+		}, {
+			desc:       pDesc,
+			lowID:      inId90,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   false,
+			wantWrite:  false,
+			wantRead:   true,
+			wantStatus: 6,
+		},
+	}
+	var p4Clients []*p4rt_client.P4RTClient
+	for _, test := range testCases {
+		t.Run(test.desc, func(t *testing.T) {
+			resp, err := streamP4RTArb(&test)
+			if err != nil && test.wantFail {
+				t.Errorf("Failed to setup P4RT Client: %v", err)
+			}
+			// Validate status code
+			if resp != test.wantStatus {
+				t.Errorf("Incorrect status code received: want %d, got %d", test.wantStatus, resp)
+			}
+			// Validate the response for Read/Write for each client
+			validateRWResp(t, &test)
+			// Keep track of connections for teardown
+			p4Clients = append(p4Clients, test.handle)
+		})
+	}
+	// Teardown clients
+	for _, p4Client := range p4Clients {
+		removeClient(p4Client)
+	}
+}
+
+// Test Duplicate ElectionID
+func TestDuplicateElectionID(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	configureDeviceId(t, dut)
+	configurePortId(t, dut)
+	testCases := []testArgs{
+		{
+			desc:       pDesc,
+			lowID:      inId100,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   false,
+			wantStatus: 0,
+		}, {
+			desc:       sDesc,
+			lowID:      inId100,
+			highID:     inId0,
+			handle:     clientConnection(t, dut),
+			deviceID:   deviceId,
+			wantFail:   true,
+			wantStatus: 0,
```

Comment:
Should it be 3?

```
If the election_id is set and is already used by another controller (excluding the controller making the request) for the same (device_id, role), the P4Runtime server shall terminate the stream by returning an INVALID_ARGUMENT error.
```

error code for INVALID_ARGUMENTis 3

Reply:
For the below Testcase, sometimes we are observing that the testcase did not receive the status code 3, as the P4RT server already disconnects the streamChannel after sending the MasterArbitration Response. 

 func TestDuplicateElectionID(t *testing.T) { 

Reference from the specs (5.3.2.4)
If the election_id is set and is already used by another controller (excluding the controller making the request) for the same (device_id, role), the P4Runtime server shall terminate the stream by returning an INVALID_ARGUMENT error.

Reply:
> In the below testcase(func TestPrimaryReconnect(t *testing.T)), the client is getting disconnected after every iteration. This means the connection should be successful for each iteration with election ID 100, 90 and 100 again. Since there is only one client exists within a given iteration, for iteration 2, the wantStatus should be 0 instead of 5?
> 
> https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L327

I think there might be a disconnect here, the server would keep track of the Primary Election ID, should a connection come in with anything < Primary Election ID - Even if the Primary is disconnected, the error MSG should be 5.

From Section 16.2 - https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
> the P4Runtime server keeps track of the highest election_id that it has ever received. If a controller's election_id is equal to the highest election_id that the server has ever received, that controller is the primary. All other controllers are backups. Note that it is possible that all controllers are backups, and that there is no primary. There can be at most one primary, because for any given (device_id, role), each connected controller has a unique election_id.

And Section 5.4: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
>If there is a primary:
For the primary, status is OK (with status.code set to google.rpc.OK).
For all backup controllers, status is set to non-OK (with status.code set to google.rpc.ALREADY_EXISTS).
Otherwise, if there is no primary currently, for all backup controllers, status is set to non-OK (with status.code set to google.rpc.NOT_FOUND).

GRPC status codes for ref: https://developers.google.com/maps-booking/reference/grpc-api/status_codes

Reply:
I'm currently looking into this case, as you mention the streamChannel gets disconnected right after the MasterArbitration Response, making it challenging to capture the Arbitration response status code. Might have to look at changing the client, but it's a WIP.

Reply:
Thanks @Hobbydos

Reply:
@Hobbydos please refer to other tests to see how to capture Stream Termination Errors and then update to the correct status codes to validate

Reply:
Thanks @Ankur19  - I plan to revisit this shortly. It seems like I'll have to change how the p4rt node name is passed, too.
Ref:

Tests should get the P4RT Node Name by walking the Components OC tree. Components of type INTEGRATED_CIRCUIT should have child Components of type PORT. These PORT Components can be mapped to currently reserved Interfaces using the hardware-port leaf in the Interfaces tree. Such an implementation already exists in p4rtutils library: p4rtutils.P4RTNodesByPort().

If P4RT Node Names cannot be resolved by walking the Components tree, use deviation flag --deviation_explicit_p4rt_node_component and pass the node names through args --arg_p4rt_node_name_1, --arg_p4rt_node_name_2.

Reply:
@Ankur19 The requested changes have been added